Importing some libraries

In [1]:
import pandas as pd

In [2]:
import numpy as np

Importing the data

In [3]:
df=pd.read_csv("/content/weatherAUS.csv")

Broader Anaysis of data

In [4]:
df.head(5)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

In [5]:
df.shape


(145460, 23)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

Before we process anything we see that more than half of the data corresponding to the "Evaporation","Sunshine","Cloud9am","Cloud3pm" features are missing and we already have and will get more features by breaking , encoding other features later on. So it is better to drop whole of these features now early on.

In [7]:
df=df.drop(["Evaporation","Sunshine","Cloud9am","Cloud3pm"],axis=1)
df

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0       2008-12-01   Albury     13.4     22.9       0.6           W   
1       2008-12-02   Albury      7.4     25.1       0.0         WNW   
2       2008-12-03   Albury     12.9     25.7       0.0         WSW   
3       2008-12-04   Albury      9.2     28.0       0.0          NE   
4       2008-12-05   Albury     17.5     32.3       1.0           W   
...            ...      ...      ...      ...       ...         ...   
145455  2017-06-21    Uluru      2.8     23.4       0.0           E   
145456  2017-06-22    Uluru      3.6     25.3       0.0         NNW   
145457  2017-06-23    Uluru      5.4     26.9       0.0           N   
145458  2017-06-24    Uluru      7.8     27.0       0.0          SE   
145459  2017-06-25    Uluru     14.9      NaN       0.0         NaN   

        WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0                44.0          W        WNW          20.0          24.0   
1                44.0        NNW        WSW           4.0          22.0   
2                46.0          W        WSW          19.0          26.0   
3                24.0         SE          E          11.0           9.0   
4                41.0        ENE         NW           7.0          20.0   
...               ...        ...        ...           ...           ...   
145455           31.0         SE        ENE          13.0          11.0   
145456           22.0         SE          N          13.0           9.0   
145457           37.0         SE        WNW           9.0           9.0   
145458           28.0        SSE          N          13.0           7.0   
145459            NaN        ESE        ESE          17.0          17.0   

        Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
0              71.0         22.0       1007.7       1007.1     16.9     21.8   
1              44.0         25.0       1010.6       1007.8     17.2     24.3   
2              38.0         30.0       1007.6       1008.7     21.0     23.2   
3              45.0         16.0       1017.6       1012.8     18.1     26.5   
4              82.0         33.0       1010.8       1006.0     17.8     29.7   
...             ...          ...          ...          ...      ...      ...   
145455         51.0         24.0       1024.6       1020.3     10.1     22.4   
145456         56.0         21.0       1023.5       1019.1     10.9     24.5   
145457         53.0         24.0       1021.0       1016.8     12.5     26.1   
145458         51.0         24.0       1019.4       1016.5     15.1     26.0   
145459         62.0         36.0       1020.2       1017.9     15.0     20.9   

       RainToday RainTomorrow  
0             No           No  
1             No           No  
2             No           No  
3             No           No  
4             No           No  
...          ...          ...  
145455        No           No  
145456        No           No  
145457        No           No  
145458        No           No  
145459        No          NaN  

[145460 rows x 19 columns]

Here are some missing values in other features as well , we can simply remove the corresponding rows because we have a large amount of data present.

In [8]:
df.dropna(inplace=True)
df

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  \
0       2008-12-01   Albury     13.4     22.9       0.6           W   
1       2008-12-02   Albury      7.4     25.1       0.0         WNW   
2       2008-12-03   Albury     12.9     25.7       0.0         WSW   
3       2008-12-04   Albury      9.2     28.0       0.0          NE   
4       2008-12-05   Albury     17.5     32.3       1.0           W   
...            ...      ...      ...      ...       ...         ...   
145454  2017-06-20    Uluru      3.5     21.8       0.0           E   
145455  2017-06-21    Uluru      2.8     23.4       0.0           E   
145456  2017-06-22    Uluru      3.6     25.3       0.0         NNW   
145457  2017-06-23    Uluru      5.4     26.9       0.0           N   
145458  2017-06-24    Uluru      7.8     27.0       0.0          SE   

        WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  \
0                44.0          W        WNW          20.0          24.0   
1                44.0        NNW        WSW           4.0          22.0   
2                46.0          W        WSW          19.0          26.0   
3                24.0         SE          E          11.0           9.0   
4                41.0        ENE         NW           7.0          20.0   
...               ...        ...        ...           ...           ...   
145454           31.0        ESE          E          15.0          13.0   
145455           31.0         SE        ENE          13.0          11.0   
145456           22.0         SE          N          13.0           9.0   
145457           37.0         SE        WNW           9.0           9.0   
145458           28.0        SSE          N          13.0           7.0   

        Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  \
0              71.0         22.0       1007.7       1007.1     16.9     21.8   
1              44.0         25.0       1010.6       1007.8     17.2     24.3   
2              38.0         30.0       1007.6       1008.7     21.0     23.2   
3              45.0         16.0       1017.6       1012.8     18.1     26.5   
4              82.0         33.0       1010.8       1006.0     17.8     29.7   
...             ...          ...          ...          ...      ...      ...   
145454         59.0         27.0       1024.7       1021.2      9.4     20.9   
145455         51.0         24.0       1024.6       1020.3     10.1     22.4   
145456         56.0         21.0       1023.5       1019.1     10.9     24.5   
145457         53.0         24.0       1021.0       1016.8     12.5     26.1   
145458         51.0         24.0       1019.4       1016.5     15.1     26.0   

       RainToday RainTomorrow  
0             No           No  
1             No           No  
2             No           No  
3             No           No  
4             No           No  
...          ...          ...  
145454        No           No  
145455        No           No  
145456        No           No  
145457        No           No  
145458        No           No  

[112925 rows x 19 columns]

We also need to make sure our data does not has repeated rows.

In [9]:
df.duplicated().value_counts()

False    112925
dtype: int64

This means our data was safe from repetition.

Analysing Date Feature

In [10]:
df["Date"].head()

0    2008-12-01
1    2008-12-02
2    2008-12-03
3    2008-12-04
4    2008-12-05
Name: Date, dtype: object

As we can see dates is present as object so we need to convert it into datatime object so we can futher process it and break it into more useful features .

In [11]:
df["Date"] = pd.to_datetime(df["Date"],
 format = "%Y-%m-%d",
 errors = "coerce")

In [12]:
df.Date.head()

0   2008-12-01
1   2008-12-02
2   2008-12-03
3   2008-12-04
4   2008-12-05
Name: Date, dtype: datetime64[ns]

Breaking the datatime object into some useful features such as day , month , year

In [13]:
df["Day"]=df.Date.dt.day
df["Month"]=df.Date.dt.month
df["Year"]=df.Date.dt.year

We can now see our new features in the end of dataframe.

In [14]:
df.head()

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0 2008-12-01   Albury     13.4     22.9       0.6           W           44.0   
1 2008-12-02   Albury      7.4     25.1       0.0         WNW           44.0   
2 2008-12-03   Albury     12.9     25.7       0.0         WSW           46.0   
3 2008-12-04   Albury      9.2     28.0       0.0          NE           24.0   
4 2008-12-05   Albury     17.5     32.3       1.0           W           41.0   

  WindDir9am WindDir3pm  WindSpeed9am  ...  Humidity3pm  Pressure9am  \
0          W        WNW          20.0  ...         22.0       1007.7   
1        NNW        WSW           4.0  ...         25.0       1010.6   
2          W        WSW          19.0  ...         30.0       1007.6   
3         SE          E          11.0  ...         16.0       1017.6   
4        ENE         NW           7.0  ...         33.0       1010.8   

   Pressure3pm  Temp9am  Temp3pm  RainToday  RainTomorrow Day Month  Year  
0       1007.1     16.9     21.8         No            No   1    12  2008  
1       1007.8     17.2     24.3         No            No   2    12  2008  
2       1008.7     21.0     23.2         No            No   3    12  2008  
3       1012.8     18.1     26.5         No            No   4    12  2008  
4       1006.0     17.8     29.7         No            No   5    12  2008  

[5 rows x 22 columns]

Now we don't need "Date" since we already have its replacements so we can drop it.

In [15]:
df=df.drop(["Date"],axis=1)
df.head()

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed WindDir9am  \
0   Albury     13.4     22.9       0.6           W           44.0          W   
1   Albury      7.4     25.1       0.0         WNW           44.0        NNW   
2   Albury     12.9     25.7       0.0         WSW           46.0          W   
3   Albury      9.2     28.0       0.0          NE           24.0         SE   
4   Albury     17.5     32.3       1.0           W           41.0        ENE   

  WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  Pressure9am  \
0        WNW          20.0          24.0  ...         22.0       1007.7   
1        WSW           4.0          22.0  ...         25.0       1010.6   
2        WSW          19.0          26.0  ...         30.0       1007.6   
3          E          11.0           9.0  ...         16.0       1017.6   
4         NW           7.0          20.0  ...         33.0       1010.8   

   Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow Day  Month  Year  
0       1007.1     16.9     21.8         No           No   1     12  2008  
1       1007.8     17.2     24.3         No           No   2     12  2008  
2       1008.7     21.0     23.2         No           No   3     12  2008  
3       1012.8     18.1     26.5         No           No   4     12  2008  
4       1006.0     17.8     29.7         No           No   5     12  2008  

[5 rows x 21 columns]

Lets check some information about updated data.

In [16]:
df.describe()

MinTemp        MaxTemp       Rainfall  WindGustSpeed  \
count  112925.000000  112925.000000  112925.000000  112925.000000   
mean       12.664721      23.655670       2.377892      40.786611   
std         6.254135       6.982702       8.602968      13.321774   
min        -8.200000       2.600000       0.000000       7.000000   
25%         8.100000      18.300000       0.000000      31.000000   
50%        12.400000      23.100000       0.000000      39.000000   
75%        17.200000      28.700000       0.800000      48.000000   
max        33.900000      48.100000     367.600000     135.000000   

        WindSpeed9am   WindSpeed3pm    Humidity9am    Humidity3pm  \
count  112925.000000  112925.000000  112925.000000  112925.000000   
mean       15.179163      19.500545      67.404162      50.669055   
std         8.344304       8.580097      18.911610      20.768550   
min         2.000000       2.000000       0.000000       0.000000   
25%         9.000000      13.000000      56.000000      36.000000   
50%        13.000000      19.000000      68.000000      51.000000   
75%        20.000000      24.000000      81.000000      65.000000   
max        87.000000      87.000000     100.000000     100.000000   

         Pressure9am    Pressure3pm        Temp9am        Temp3pm  \
count  112925.000000  112925.000000  112925.000000  112925.000000   
mean     1017.425847    1015.048962      17.462008      22.127306   
std         7.032654       6.957822       6.355045       6.837380   
min       980.500000     977.100000      -3.100000       1.700000   
25%      1012.800000    1010.300000      12.700000      16.900000   
50%      1017.400000    1015.000000      17.100000      21.600000   
75%      1022.100000    1019.700000      22.000000      26.900000   
max      1041.000000    1039.600000      40.200000      46.700000   

                 Day          Month           Year  
count  112925.000000  112925.000000  112925.000000  
mean       15.714678       6.431499    2012.753385  
std         8.790323       3.476314       2.530203  
min         1.000000       1.000000    2007.000000  
25%         8.000000       3.000000    2011.000000  
50%        16.000000       6.000000    2013.000000  
75%        23.000000       9.000000    2015.000000  
max        31.000000      12.000000    2017.000000

We see our data especially "Rainfall" , "Evaporation","WindSpeed9am" are having outliners definitely as can be seen from the difference between the 75th percentile data comparing to the max values. Even the other way round "Humidity9am" seems to have significant difference between the 25th percentile and min value.

So , we should remove rows corresponding to the outliners.

In [17]:
dfx=df.describe()
def remove_outliners():
  for i in dfx.columns:
    # if(i=="Day" or i=="Month" or i=="Year"):
        # continue;

    print(i,dfx[i]["mean"])
    print("z for",i,3*dfx[i]["std"]+dfx[i]["mean"])
    df.drop((df[df[i]>3*dfx[i]["std"]+dfx[i]["mean"]]).index,inplace=True,axis=0)
    # mean=dfx[i]["mean"];
    # std=dfx[i]["std"]


remove_outliners()

MinTemp 12.664720832410893
z for MinTemp 31.427124433331404
MaxTemp 23.655669692273634
z for MaxTemp 44.60377714601473
Rainfall 2.3778924064644675
z for Rainfall 28.186796589605535
WindGustSpeed 40.786610582244855
z for WindGustSpeed 80.75193137200424
WindSpeed9am 15.179163161390303
z for WindSpeed9am 40.21207453959913
WindSpeed3pm 19.50054460925393
z for WindSpeed3pm 45.24083699885332
Humidity9am 67.40416205446093
z for Humidity9am 124.13899147975329
Humidity3pm 50.66905468231127
z for Humidity3pm 112.97470441743104
Pressure9am 1017.4258472437458
z for Pressure9am 1038.5238094022134
Pressure3pm 1015.0489617002436
z for Pressure3pm 1035.9224275845777
Temp9am 17.46200752711977
z for Temp9am 36.52714168410968
Temp3pm 22.12730573389418
z for Temp3pm 42.63944596712517
Day 15.71467788355103
z for Day 42.08564757921309
Month 6.431498782377684
z for Month 16.86044089896551
Year 2012.7533849900376
z for Year 2020.3439928826258


In [18]:
df

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0        Albury     13.4     22.9       0.6           W           44.0   
1        Albury      7.4     25.1       0.0         WNW           44.0   
2        Albury     12.9     25.7       0.0         WSW           46.0   
3        Albury      9.2     28.0       0.0          NE           24.0   
4        Albury     17.5     32.3       1.0           W           41.0   
...         ...      ...      ...       ...         ...            ...   
145454    Uluru      3.5     21.8       0.0           E           31.0   
145455    Uluru      2.8     23.4       0.0           E           31.0   
145456    Uluru      3.6     25.3       0.0         NNW           22.0   
145457    Uluru      5.4     26.9       0.0           N           37.0   
145458    Uluru      7.8     27.0       0.0          SE           28.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  \
0               W        WNW          20.0          24.0  ...         22.0   
1             NNW        WSW           4.0          22.0  ...         25.0   
2               W        WSW          19.0          26.0  ...         30.0   
3              SE          E          11.0           9.0  ...         16.0   
4             ENE         NW           7.0          20.0  ...         33.0   
...           ...        ...           ...           ...  ...          ...   
145454        ESE          E          15.0          13.0  ...         27.0   
145455         SE        ENE          13.0          11.0  ...         24.0   
145456         SE          N          13.0           9.0  ...         21.0   
145457         SE        WNW           9.0           9.0  ...         24.0   
145458        SSE          N          13.0           7.0  ...         24.0   

        Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  \
0            1007.7       1007.1     16.9     21.8         No           No   
1            1010.6       1007.8     17.2     24.3         No           No   
2            1007.6       1008.7     21.0     23.2         No           No   
3            1017.6       1012.8     18.1     26.5         No           No   
4            1010.8       1006.0     17.8     29.7         No           No   
...             ...          ...      ...      ...        ...          ...   
145454       1024.7       1021.2      9.4     20.9         No           No   
145455       1024.6       1020.3     10.1     22.4         No           No   
145456       1023.5       1019.1     10.9     24.5         No           No   
145457       1021.0       1016.8     12.5     26.1         No           No   
145458       1019.4       1016.5     15.1     26.0         No           No   

       Day  Month  Year  
0        1     12  2008  
1        2     12  2008  
2        3     12  2008  
3        4     12  2008  
4        5     12  2008  
...     ..    ...   ...  
145454  20      6  2017  
145455  21      6  2017  
145456  22      6  2017  
145457  23      6  2017  
145458  24      6  2017  

[108552 rows x 21 columns]

Analysing the features one by one

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108552 entries, 0 to 145458
Data columns (total 21 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Location       108552 non-null  object 
 1   MinTemp        108552 non-null  float64
 2   MaxTemp        108552 non-null  float64
 3   Rainfall       108552 non-null  float64
 4   WindGustDir    108552 non-null  object 
 5   WindGustSpeed  108552 non-null  float64
 6   WindDir9am     108552 non-null  object 
 7   WindDir3pm     108552 non-null  object 
 8   WindSpeed9am   108552 non-null  float64
 9   WindSpeed3pm   108552 non-null  float64
 10  Humidity9am    108552 non-null  float64
 11  Humidity3pm    108552 non-null  float64
 12  Pressure9am    108552 non-null  float64
 13  Pressure3pm    108552 non-null  float64
 14  Temp9am        108552 non-null  float64
 15  Temp3pm        108552 non-null  float64
 16  RainToday      108552 non-null  object 
 17  RainTomorrow   108552 non-nul

In [20]:
pd.DataFrame(df["RainToday"].value_counts())

RainToday
No       85865
Yes      22687

In [21]:
pd.DataFrame(df["RainTomorrow"].value_counts())

RainTomorrow
No          85888
Yes         22664

In [22]:
dfresult=pd.DataFrame([df["RainTomorrow"].value_counts(),df["RainToday"].value_counts()])
dfresult

No    Yes
RainTomorrow  85888  22664
RainToday     85865  22687

There is huge difference in "Yes" and "No" for RainTomorrow. If we want we can try sampling out data as well but that will reduce data size considerably so lets leave this for now.

The numbers of "RainToday" and "RainTomorrow" are quite similar.

In [23]:
dfx=df[(df["RainToday"]=="Yes") & (df["RainTomorrow"]=="Yes")]
dfx

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
11       Albury     15.9     21.7       2.2         NNE           31.0   
12       Albury     15.9     18.6      15.6           W           61.0   
17       Albury     13.5     22.9      16.8           W           63.0   
103      Albury     16.4     27.0       3.0          SW           69.0   
131      Albury     13.0     24.9       8.4           W           46.0   
...         ...      ...      ...       ...         ...            ...   
145305    Uluru     24.5     37.5       3.2         NNE           48.0   
145306    Uluru     23.1     34.5       6.4         SSW           50.0   
145321    Uluru     24.4     37.9       7.6          NE           72.0   
145391    Uluru     15.2     21.5       6.8         ENE           30.0   
145392    Uluru     17.7     26.9      12.6           S           35.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  \
11             NE        ENE          15.0          13.0  ...         91.0   
12            NNW        NNW          28.0          28.0  ...         93.0   
17              N        WNW           6.0          20.0  ...         65.0   
103             E          N           9.0          22.0  ...         74.0   
131             S          E           6.0          11.0  ...         47.0   
...           ...        ...           ...           ...  ...          ...   
145305        ENE        WSW           4.0          13.0  ...         38.0   
145306        ENE          W          19.0           7.0  ...         56.0   
145321          E         NE          17.0          31.0  ...         35.0   
145391         NE        ENE          19.0           9.0  ...         65.0   
145392          E        SSW          15.0          20.0  ...         59.0   

        Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  \
11           1010.5       1004.2     15.9     17.0        Yes          Yes   
12            994.3        993.0     17.4     15.8        Yes          Yes   
17           1005.8       1002.2     18.0     21.5        Yes          Yes   
103          1012.7       1008.0     19.9     19.7        Yes          Yes   
131          1024.4       1020.3     18.0     24.4        Yes          Yes   
...             ...          ...      ...      ...        ...          ...   
145305       1007.9       1003.7     28.4     36.1        Yes          Yes   
145306       1007.8       1003.7     26.0     32.9        Yes          Yes   
145321       1006.2       1002.2     27.7     37.0        Yes          Yes   
145391       1016.9       1015.3     19.0     21.2        Yes          Yes   
145392       1018.1       1014.7     19.0     26.0        Yes          Yes   

       Day  Month  Year  
11      12     12  2008  
12      13     12  2008  
17      18     12  2008  
103     14      3  2009  
131     11      4  2009  
...     ..    ...   ...  
145305  22      1  2017  
145306  23      1  2017  
145321   7      2  2017  
145391  18      4  2017  
145392  19      4  2017  

[9989 rows x 21 columns]

So, nearly (little less) half of the time if it rains today there is the possibility of raining tomorrow .

Analysing "Location" feature

In [26]:
df["Location"].value_counts()

Perth               3004
Darwin              2961
Brisbane            2939
Hobart              2884
Mildura             2858
PerthAirport        2856
MountGambier        2809
Townsville          2772
Cobar               2766
Woomera             2760
NorfolkIsland       2750
WaggaWagga          2746
Nuriootpa           2740
Cairns              2727
Ballarat            2723
Portland            2719
Adelaide            2708
AliceSprings        2701
Watsonia            2694
SydneyAirport       2690
Bendigo             2675
Canberra            2648
MelbourneAirport    2630
GoldCoast           2600
NorahHead           2589
Sale                2587
Moree               2567
Wollongong          2504
Walpole             2441
PearceRAAF          2404
CoffsHarbour        2396
Albury              2387
Tuggeranong         2269
BadgerysCreek       2264
Witchcliffe         2254
Dartmoor            2247
Sydney              2158
Williamtown         2064
Melbourne           2014
Richmond            1983


The distribution is ok. (not the case of very low or high single location)

If we want we can remove the "Katherine" rows.

In [27]:
df.drop(df[df["Location"]=="Katherine"].index,axis=0)

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0        Albury     13.4     22.9       0.6           W           44.0   
1        Albury      7.4     25.1       0.0         WNW           44.0   
2        Albury     12.9     25.7       0.0         WSW           46.0   
3        Albury      9.2     28.0       0.0          NE           24.0   
4        Albury     17.5     32.3       1.0           W           41.0   
...         ...      ...      ...       ...         ...            ...   
145454    Uluru      3.5     21.8       0.0           E           31.0   
145455    Uluru      2.8     23.4       0.0           E           31.0   
145456    Uluru      3.6     25.3       0.0         NNW           22.0   
145457    Uluru      5.4     26.9       0.0           N           37.0   
145458    Uluru      7.8     27.0       0.0          SE           28.0   

       WindDir9am WindDir3pm  WindSpeed9am  WindSpeed3pm  ...  Humidity3pm  \
0               W        WNW          20.0          24.0  ...         22.0   
1             NNW        WSW           4.0          22.0  ...         25.0   
2               W        WSW          19.0          26.0  ...         30.0   
3              SE          E          11.0           9.0  ...         16.0   
4             ENE         NW           7.0          20.0  ...         33.0   
...           ...        ...           ...           ...  ...          ...   
145454        ESE          E          15.0          13.0  ...         27.0   
145455         SE        ENE          13.0          11.0  ...         24.0   
145456         SE          N          13.0           9.0  ...         21.0   
145457         SE        WNW           9.0           9.0  ...         24.0   
145458        SSE          N          13.0           7.0  ...         24.0   

        Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  \
0            1007.7       1007.1     16.9     21.8         No           No   
1            1010.6       1007.8     17.2     24.3         No           No   
2            1007.6       1008.7     21.0     23.2         No           No   
3            1017.6       1012.8     18.1     26.5         No           No   
4            1010.8       1006.0     17.8     29.7         No           No   
...             ...          ...      ...      ...        ...          ...   
145454       1024.7       1021.2      9.4     20.9         No           No   
145455       1024.6       1020.3     10.1     22.4         No           No   
145456       1023.5       1019.1     10.9     24.5         No           No   
145457       1021.0       1016.8     12.5     26.1         No           No   
145458       1019.4       1016.5     15.1     26.0         No           No   

       Day  Month  Year  
0        1     12  2008  
1        2     12  2008  
2        3     12  2008  
3        4     12  2008  
4        5     12  2008  
...     ..    ...   ...  
145454  20      6  2017  
145455  21      6  2017  
145456  22      6  2017  
145457  23      6  2017  
145458  24      6  2017  

[107905 rows x 21 columns]

Analysing "WindDir9am" and "WindDir3pm".

In [28]:
dfresult=pd.DataFrame([df["WindDir9am"].value_counts(),df["WindDir3pm"].value_counts()])
dfresult

N   SSE     E    SE     S   ENE   NNE    SW   ESE   NNW    NW  \
WindDir9am  9083  7841  7811  7598  7160  6727  6721  6706  6482  6304  6297   
WindDir3pm  6797  7435  6548  8065  7801  6371  5137  7433  6687  5824  6012   

               W    NE   SSW   WNW   WSW  
WindDir9am  6279  6239  5995  5671  5638  
WindDir3pm  7518  6474  6420  6542  7488

The distribution here is also fine.

Analysing "WindGustDir"

In [29]:
pd.DataFrame([df["WindGustDir"].value_counts()])

W    SE     E   SSE    SW     S   WSW     N   SSW   ENE   WNW  \
WindGustDir  7957  7789  7649  7423  7373  7342  7322  7182  7112  6612  6322   

               NW   ESE    NE   NNE   NNW  
WindGustDir  6064  6056  5930  5360  5059

Encoding the data to make the categorical features as numerical so that we can train it in our model.

In [30]:
encoded_data=pd.get_dummies(df,columns=['WindDir9am','WindDir3pm',"Location","WindGustDir","RainToday","RainTomorrow"],drop_first=True)

In [31]:
encoded_data.head()

MinTemp  MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  WindSpeed3pm  \
0     13.4     22.9       0.6           44.0          20.0          24.0   
1      7.4     25.1       0.0           44.0           4.0          22.0   
2     12.9     25.7       0.0           46.0          19.0          26.0   
3      9.2     28.0       0.0           24.0          11.0           9.0   
4     17.5     32.3       1.0           41.0           7.0          20.0   

   Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  ...  WindGustDir_S  \
0         71.0         22.0       1007.7       1007.1  ...              0   
1         44.0         25.0       1010.6       1007.8  ...              0   
2         38.0         30.0       1007.6       1008.7  ...              0   
3         45.0         16.0       1017.6       1012.8  ...              0   
4         82.0         33.0       1010.8       1006.0  ...              0   

   WindGustDir_SE  WindGustDir_SSE  WindGustDir_SSW  WindGustDir_SW  \
0               0                0                0               0   
1               0                0                0               0   
2               0                0                0               0   
3               0                0                0               0   
4               0                0                0               0   

   WindGustDir_W  WindGustDir_WNW  WindGustDir_WSW  RainToday_Yes  \
0              1                0                0              0   
1              0                1                0              0   
2              0                0                1              0   
3              0                0                0              0   
4              1                0                0              0   

   RainTomorrow_Yes  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 105 columns]

Separating the input (x) and output (y).

In [32]:
x=encoded_data.drop(["RainTomorrow_Yes"],axis=1)

In [33]:
x.head()

MinTemp  MaxTemp  Rainfall  WindGustSpeed  WindSpeed9am  WindSpeed3pm  \
0     13.4     22.9       0.6           44.0          20.0          24.0   
1      7.4     25.1       0.0           44.0           4.0          22.0   
2     12.9     25.7       0.0           46.0          19.0          26.0   
3      9.2     28.0       0.0           24.0          11.0           9.0   
4     17.5     32.3       1.0           41.0           7.0          20.0   

   Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  ...  WindGustDir_NW  \
0         71.0         22.0       1007.7       1007.1  ...               0   
1         44.0         25.0       1010.6       1007.8  ...               0   
2         38.0         30.0       1007.6       1008.7  ...               0   
3         45.0         16.0       1017.6       1012.8  ...               0   
4         82.0         33.0       1010.8       1006.0  ...               0   

   WindGustDir_S  WindGustDir_SE  WindGustDir_SSE  WindGustDir_SSW  \
0              0               0                0                0   
1              0               0                0                0   
2              0               0                0                0   
3              0               0                0                0   
4              0               0                0                0   

   WindGustDir_SW  WindGustDir_W  WindGustDir_WNW  WindGustDir_WSW  \
0               0              1                0                0   
1               0              0                1                0   
2               0              0                0                1   
3               0              0                0                0   
4               0              1                0                0   

   RainToday_Yes  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 104 columns]

In [34]:
y=encoded_data["RainTomorrow_Yes"]

In [35]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: RainTomorrow_Yes, dtype: uint8

In [36]:
y.value_counts()

0    85888
1    22664
Name: RainTomorrow_Yes, dtype: int64

Spliting the data into training and test data.

In [37]:
from sklearn.model_selection import train_test_split

In [38]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,test_size=0.3,random_state=42)

Since the differences between the values are big i shall standardize the data as well.

In [44]:
from sklearn.preprocessing import StandardScaler

In [40]:
st=StandardScaler()

In [41]:
xtrain=st.fit_transform(xtrain)

In [42]:
xtest=st.transform(xtest)

Accuracy Measures

In [45]:
from sklearn.metrics import accuracy_score

Model-Decision Tree

In [46]:
from sklearn.tree import DecisionTreeClassifier

In [47]:
dtc=DecisionTreeClassifier()

In [48]:
dtc.fit(xtrain,ytrain)

DecisionTreeClassifier()

In [49]:
y_dtc_predict=dtc.predict(xtest)

In [50]:
accuracy_score(y_dtc_predict,ytest)

0.7961370754774919

SVM

In [56]:
from sklearn.svm import SVC

In [57]:
svcmodel=SVC()

In [58]:
svcmodel.fit(xtrain,ytrain)

SVC()

In [59]:
y_svc_predict=svcmodel.predict(xtest)

In [60]:
accuracy_score(y_svc_predict,ytest)

0.866762881532887

RandomForestClassifier

In [51]:
from sklearn.ensemble import RandomForestClassifier

In [52]:
rfcmodel=RandomForestClassifier()

In [53]:
rfcmodel.fit(xtrain,ytrain)

RandomForestClassifier()

In [54]:
y_rfcmodel_predict=rfcmodel.predict(xtest)

In [55]:
accuracy_score(y_rfcmodel_predict,ytest)

0.8613277651538415

Logistic Regression Model

In [61]:
from sklearn import linear_model

In [62]:
logr = linear_model.LogisticRegression()
logr.fit(x,y)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [63]:
y_logrmodel_predict=rfcmodel.predict(xtest)

In [64]:
accuracy_score(y_logrmodel_predict,ytest)

0.8613277651538415

So, finally the ML model has attained near 86% accuracy maximum

SVC , RandomForestClassifier and LogisticRegression have almost similar accuracy and since SVM is slower on computation other two methods are better.